### PR TITLE
Simplify parser declarations when specialist types are stored in strings

### PR DESCRIPTION
**Pull request by przemekwitek**
This PR:
- implements `AbstractObjectParser.declareStringField` method
- makes use of it in `ml` code

This saves some LoC as the new way of declaring parsed fields is more concise.

### CODE REVIEW

**Comment by przemekwitek** [comment]
run elasticsearch-ci/2

**Review by droberts195 — APPROVED** [review]
LGTM

This looks like a great change for removing repeated boilerplate.

Before merging I think someone who knows more about the parser code should have a look too, just in case the name of the new method in `AbstractObjectParser` doesn't fit some pattern.  Maybe @nik9000?

**Comment by przemekwitek** [comment]
> LGTM
> 
> This looks like a great change for removing repeated boilerplate.
> 
> Before merging I think someone who knows more about the parser code should have a look too, just in case the name of the new method in `AbstractObjectParser` doesn't fit some pattern. Maybe @nik9000?

Thanks for suggestion. I've added Nik as a reviewer.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java:234]: I *thought* that this wouldn't get called unless it is on a string because of the `ValueType.STRING`.

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java:230]: Would you be ok renaming to `declareString`?

File: libs/x-content/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java
Comment (nik9000) [inline libs/x-content/src/test/java/org/elasticsearch/common/xcontent/ObjectParserTests.java:341]: I'd prefer not to have an _ in the method name.

**Review by nik9000 — COMMENTED** [review]
I like it! I think there are a couple ways to do this but we do it so frequently that having a helper is nice.

I left three small things worth talking about before merging, but I'm excited to have this!

**Comment by nik9000** [comment]
I'm cool with removing the if block. I think it comes from the guts of
ObjectParser. Lgtm.

**Pull request merged**
